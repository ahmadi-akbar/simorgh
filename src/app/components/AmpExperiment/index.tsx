/** @jsx jsx */
/* @jsxFrag React.Fragment */
import { jsx } from '@emotion/react';
import React from 'react';
import { Helmet } from 'react-helmet';

type AmpExperimentConfig = {
  [key: string]: {
    sticky?: boolean;
    consentNotificationId?: string;
    variants: {
      [key: string]: number; // variantName: traffic allocation %
    };
  };
};

type AmpExperiment = {
  [key: string]: AmpExperimentConfig;
};

const AmpHead = () => (
  <Helmet>
    <script
      async
      custom-element="amp-experiment"
      src="https://cdn.ampproject.org/v0/amp-experiment-0.1.js"
    />
  </Helmet>
);

const AmpExperiment = ({ experimentData }: AmpExperiment) => {
  return (
    <>
      <AmpHead />
      <amp-experiment>
        <script
          type="application/json"
          /* eslint-disable-next-line react/no-danger */
          dangerouslySetInnerHTML={{ __html: JSON.stringify(experimentData) }}
        />
      </amp-experiment>
    </>
  );
};

export default AmpExperiment;
