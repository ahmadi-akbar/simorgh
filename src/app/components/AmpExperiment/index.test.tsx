import React from 'react';
import { render, waitFor } from '../react-testing-library-with-providers';
import AmpExperiment from './index';

const experimentConfig = {
  someExperiment: {
    variants: {
      control: 33,
      variant_1: 33,
      variant_2: 33,
    },
  },
};

const multipleExperimentConfig = {
  aExperiment: {
    variants: {
      control: 33,
      variant_1: 33,
      variant_2: 33,
    },
  },
  bExperiment: {
    variants: {
      control: 33,
      variant_1: 33,
      variant_2: 33,
    },
  },
};

describe('Amp experiment container on Amp pages', () => {
  it('should render an amp-experiment with the expected config', async () => {
    const { container } = render(
      <AmpExperiment experimentData={experimentConfig} />,
    );
    expect(container.querySelector('amp-experiment')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should render an amp-experiment with the expected config when multiple experiments are running at the same time', async () => {
    const { container } = render(
      <AmpExperiment experimentData={multipleExperimentConfig} />,
    );
    expect(container.querySelector('amp-experiment')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it(`should add amp-experiment extension script to page head`, async () => {
    render(<AmpExperiment experimentData={experimentConfig} />);

    await waitFor(() => {
      const scripts = Array.from(document.querySelectorAll('head script'));

      expect(scripts).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            src: `https://cdn.ampproject.org/v0/amp-experiment-0.1.js`,
          }),
        ]),
      );

      expect(scripts).toHaveLength(1);
    });
  });
});
