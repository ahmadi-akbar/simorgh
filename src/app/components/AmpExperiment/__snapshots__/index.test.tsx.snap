// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amp experiment container on Amp pages should render an amp-experiment with the expected config 1`] = `
<div>
  <amp-experiment>
    <script
      type="application/json"
    >
      {"someExperiment":{"variants":{"control":33,"variant_1":33,"variant_2":33}}}
    </script>
  </amp-experiment>
</div>
`;

exports[`Amp experiment container on Amp pages should render an amp-experiment with the expected config when multiple experiments are running at the same time 1`] = `
<div>
  <amp-experiment>
    <script
      type="application/json"
    >
      {"aExperiment":{"variants":{"control":33,"variant_1":33,"variant_2":33}},"bExperiment":{"variants":{"control":33,"variant_1":33,"variant_2":33}}}
    </script>
  </amp-experiment>
</div>
`;
